src/audio/buffer.c

```diff
@@ -51,23 +52,25 @@ struct comp_buffer *buffer_alloc(uint32_t size, uint32_t caps, uint32_t flags, u
 		return NULL;
 	}
 
-	buffer->stream.addr = rballoc_align(0, caps, size, align);
-	if (!buffer->stream.addr) {
+	stream_addr = rballoc_align(0, caps, size, align);
+	if (!stream_addr) {
 		rfree(buffer);
 		tr_err(&buffer_tr, "buffer_alloc(): could not alloc size = %u bytes of type = %u",
 		       size, caps);
 		return NULL;
 	}
 
-	buffer->stream.underrun_permitted = !!(flags & SOF_BUF_UNDERRUN_PERMITTED);
-	buffer->stream.overrun_permitted = !!(flags & SOF_BUF_OVERRUN_PERMITTED);
-
-	list_init(&buffer->source_list);
-	list_init(&buffer->sink_list);
-
 	/* From here no more uncached access to the buffer object, except its list headers */
 	buffer_c = buffer_acquire(buffer);
+	buffer_c->stream.addr = stream_addr;
 	buffer_init(buffer_c, size, caps);
+
+	buffer_c->stream.underrun_permitted = !!(flags & SOF_BUF_UNDERRUN_PERMITTED);
+	buffer_c->stream.overrun_permitted = !!(flags & SOF_BUF_OVERRUN_PERMITTED);
+
+	list_init(&buffer_c->source_list);
+	list_init(&buffer_c->sink_list);
```

Comment:
now these two lines kind of break the concept. On the one hand they're useless, because those list objects are *members* in lists, not list *heads*, OTOH we always do that and in some places we actually use the fact that list members are "empty heads" when they aren't on any list. And here it breaks the concept because it assigns
```
head->next = head;
```
where `head` now is a cached address and those lists are always linked with uncached addresses

Reply:
@lyakh pushed these out to uncached ops at the end in V3 version